### PR TITLE
Optimize immutable.HashMap builder (inspired by the 2.13 implementation)

### PR DESCRIPTION
**Pull request by mkeskells**
The builder allows `TrieHashMap` to be mutated in place until it is visible this reduces the memory churn while building `Map`s.

### CODE REVIEW

File: src/library/scala/collection/immutable/HashMap.scala [hunk 1]
```diff
@@ -97,17 +98,6 @@ sealed class HashMap[A, +B] extends AbstractMap[A, B]
 
   protected def filter0(p: ((A, B)) => Boolean, negate: Boolean, level: Int, buffer: Array[HashMap[A, B @uV]], offset0: Int): HashMap[A, B] = null
 
-  protected def elemHashCode(key: A) = key.##
```
Comment (retronym) [inline src/library/scala/collection/immutable/HashMap.scala]: User defined element hashing was removed in 2.13 with a migration suggestion to use use wrapper classes for keys. But I don't think we can remove this in a minor release. Did you intend to include this diff in this PR?

Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala]: I didn't change the functionality her, just moved it to be in the object, as I needed the hashing from the builder

Is your concern the binary compat, or the functionality?
If the former, the methods can remain but be unused?

Comment (retronym) [reply src/library/scala/collection/immutable/HashMap.scala]: The functionality. Admittedly it does not have test coverage but it is used a few times in the wild: https://github.com/search?q=%22override+def+elemHashCode%22&type=Code

**Comment by retronym** [comment]
Echoing my comments on #8722, this change will be targetted to 2.12.12.

File: src/library/scala/collection/immutable/HashMap.scala [hunk 2]
```diff
@@ -735,4 +730,324 @@ object HashMap extends ImmutableMapFactory[HashMap] with BitOperations.Int {
...
+                assert(oBit == bc)
+                elems
+              }
+              trie.size0 = size
+              trie.elems0 = elems
+              trie.bitmap0 = bitmap
+              trie
+          }
+        case _ => hs
+      }
```
Comment (retronym) [inline src/library/scala/collection/immutable/HashMap.scala:1097]: We should call `ScalaRuntime.releaseFence` (which will itself need to be backported but made `private[scala]`) here.

File: src/library/scala/collection/immutable/HashMap.scala (hunk 1 above)
Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala]: i have restored the extension point

**Comment by mkeskells** [comment]
squashed

File: src/library/scala/collection/immutable/HashMap.scala (hunk 2 above)
Comment (retronym) [reply src/library/scala/collection/immutable/HashMap.scala:1097]: @mkeskells  I just merged https://github.com/scala/scala/pull/8779 so you could now add `s.c.i.VM.releaseFence()` to this and #8722 in the same places as 2.13.x.

**Review by retronym — APPROVED** [review]

**Pull request merged**